### PR TITLE
CI: Add golangci-lint

### LINKED ISSUE #689: Setup golangci-lint for all golang projects
As request in https://github.com/anuvu/zot/issues/206#issuecomment-885133949

### PR DESCRIPTION
**Pull request by burntcarrot**
Signed-off-by: burntcarrot <aadhav.n1@gmail.com>

Resolves #689.

### CODE REVIEW

**Comment by burntcarrot** [comment]
@Xuanwo Lint is working as expected. Can you open an issue related to the unused variables found in the `golangci-lint` check, or do you want me to comment the unused imports?

**Comment by burntcarrot** [comment]
Since the workflow doesn't show a complete traceback, I decided to run it on my local machine:

```
pkg/iowrap/pipe.go:41:2: `rerr` is unused (structcheck)
        rerr error
        ^
pkg/iowrap/pipe.go:40:2: `werr` is unused (structcheck)
        werr error
        ^
tests/service.go:11:2: `features` is unused (structcheck)
        features     ServiceFeatures
        ^
pkg/iowrap/pipe.go:112:2: S1023: redundant `return` statement (gosimple)
        return
        ^
pkg/iowrap/pipe.go:172:2: S1023: redundant `return` statement (gosimple)
        return
        ^
pkg/fswrap/httpfs.go:46:2: SA4005: ineffective assignment to field httpFileWrapper.store (staticcheck)
        h.store = nil
        ^
pkg/fswrap/httpfs.go:47:2: SA4005: ineffective assignment to field httpFileWrapper.object (staticcheck)
        h.object = nil
        ^
pkg/fswrap/httpfs.go:48:2: SA4005: ineffective assignment to field httpFileWrapper.offset (staticcheck)
        h.offset = 0
        ^
pkg/fswrap/iofs.go:147:2: SA4005: ineffective assignment to field fileWrapper.offset (staticcheck)
        o.offset = 0
        ^
pkg/fswrap/iofs.go:145:2: SA4005: ineffective assignment to field fileWrapper.store (staticcheck)
        o.store = nil
        ^
pkg/fswrap/iofs.go:146:2: SA4005: ineffective assignment to field fileWrapper.object (staticcheck)
        o.object = nil
        ^
```

Hope this helps!

**Comment by Xuanwo** [comment]
> pkg/iowrap/pipe.go:41:2: `rerr` is unused (structcheck)
>         rerr error
>         ^

OK to remove.

> pkg/iowrap/pipe.go:40:2: `werr` is unused (structcheck)
>         werr error
>         ^

OK to remove.

> tests/service.go:11:2: `features` is unused (structcheck)
>         features     ServiceFeatures
>         ^

Use it by calling `parsePairServiceNew` and assign it like https://github.com/beyondstorage/go-service-s3/blob/master/utils.go#L158-L159.

> pkg/iowrap/pipe.go:112:2: S1023: redundant `return` statement (gosimple)
>         return
>         ^

OK to remove.

> pkg/iowrap/pipe.go:172:2: S1023: redundant `return` statement (gosimple)
>         return
>         ^

OK to remove.

> pkg/fswrap/httpfs.go:46:2: SA4005: ineffective assignment to field httpFileWrapper.store (staticcheck)
>         h.store = nil
>         ^

OK to remove.

> pkg/fswrap/httpfs.go:47:2: SA4005: ineffective assignment to field httpFileWrapper.object (staticcheck)
>         h.object = nil
>         ^

OK to remove.

> pkg/fswrap/httpfs.go:48:2: SA4005: ineffective assignment to field httpFileWrapper.offset (staticcheck)
>         h.offset = 0
>         ^

OK to remove.

> pkg/fswrap/iofs.go:147:2: SA4005: ineffective assignment to field fileWrapper.offset (staticcheck)
>         o.offset = 0
>         ^

OK to remove.

> pkg/fswrap/iofs.go:145:2: SA4005: ineffective assignment to field fileWrapper.store (staticcheck)
>         o.store = nil
>         ^

OK to remove.

> pkg/fswrap/iofs.go:146:2: SA4005: ineffective assignment to field fileWrapper.object (staticcheck)
>         o.object = nil
>         ^

OK to remove.

---

All of them are easy to solve, would you be willing to open a new PR to fix them?

**Comment by burntcarrot** [comment]
Yes, I'll be happy to make a new PR. 

I don't think removing `werr` and `rerr` would be ideal, because they might be implemented in the future. I'll remove the rest of the ineffectual assignments and redundant return statements in a new PR.

File: .github/workflows/golangci-lint.yml [hunk 1]
```diff
@@ -0,0 +1,37 @@
+name: golangci-lint
+on:
```
Comment (Xuanwo) [inline .github/workflows/golangci-lint.yml:2]: How about executing `golangco-lint` on `[push, pull_request]`?

File: .github/workflows/golangci-lint.yml [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+jobs:
+  golangci:
+    name: lint
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v2
+      - name: golangci-lint
+        uses: golangci/golangci-lint-action@v2
+        with:
+          # Optional: version of golangci-lint to use in form of v1.2 or v1.2.3 or `latest` to use the latest version
```
Comment (Xuanwo) [inline .github/workflows/golangci-lint.yml:18]: How about removing all useless comments?

File: .github/workflows/golangci-lint.yml [hunk 3]
```diff
@@ -0,0 +1,37 @@
+name: golangci-lint
+on:
+  push:
+    tags:
+      - v*
+    branches:
+      - master
+  pull_request:
+jobs:
+  golangci:
```
Comment (Xuanwo) [inline .github/workflows/golangci-lint.yml:10]:
```suggestion
  golangci_lint:
```

**Comment by burntcarrot** [comment]
Can you please implement those changes as proposed above? My editor (VSCode) has been checking out to the wrong branch and due to this I've been making improper commits.

Can you also please remove the action from PR 691 if possible. I'll try to fix my development environment, apologies for the fuss created.

**Pull request merged**